File: app/src/main/java/com/guerinet/mymartlet/model/Term.java
/*
 * Copyright 2014-2017 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.model;

import android.content.Context;
import android.support.annotation.NonNull;

import org.threeten.bp.LocalDate;

import java.io.Serializable;

/**
 * One class currentTerm, consisting of a season and a year
 * @author Julien Guerinet
 * @since 1.0.0
 */
public class Term implements Serializable {
    private static final long serialVersionUID = 1L;
    /**
     * Term season
     */
    protected @Season.Type String season;
    /**
     * Term year
     */
    protected int year;

    /**
     * Moshi Constructor
     */
    protected Term() {}

    /**
     * Default Constructor
     *
     * @param season Term season
     * @param year   Term year
     */
    public Term(@Season.Type String season, int year) {
        this.season = season;
        this.year = year;
    }

    /* GETTERS */

    /**
     * @return Term season
     */
    public @Season.Type String getSeason() {
        return season;
    }

    /**
     * @return Term year
     */
    public int getYear() {
        return year;
    }

    /* HELPERS */

    /**
     * Checks if the current currentTerm is after the given currentTerm
     *
     * @param term The currentTerm to compare
     * @return True if the current currentTerm is after the given currentTerm, false otherwise
     */
    @SuppressWarnings("SimplifiableIfStatement")
    public boolean isAfter(Term term) {
        if (year > term.getYear()) {
            //Year after
            return true;
        } else if (year < term.getYear()) {
            //Year Before
            return false;
        } else {
            //Same year: check the semesters
            return Integer.valueOf(Season.getSeasonNumber(season)) >
                    Integer.valueOf(Season.getSeasonNumber(term.getSeason()));
        }
    }

    /**
     * @return Term Id, for parsing errors
     */
    public String getId() {
        return season + " " + year;
    }

    /**
     * @param context App context
     * @return String representation of the currentTerm
     */
    public String getString(Context context) {
        return Season.getString(context, season) + " " + year;
    }

    /**
     * @return The currentTerm in a format used by McGill
     */
    @Override
    public String toString() {
        return year + Season.getSeasonNumber(season);
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof Term)) {
            return false;
        }

        Term term = (Term) object;
        return season.equals(term.getSeason()) && year == term.getYear();
    }

    /* STATIC HELPERS */

    /**
     * @param term Term in String format (Ex: 199901 would be Winter 1999)
     * @return Corresponding {@link Term}
     */
    public static Term parseMcGillTerm(String term) {
        //Split it into the year and the season
        int year = Integer.parseInt(term.substring(0, 4));
        @Season.Type String season = Season.getMcGillSeason(term.substring(4));
        return new Term(season, year);
    }

    /**
     * Parses a currentTerm from a String
     *
     * @param term The currentTerm String
     * @return The parsed currentTerm
     */
    @NonNull
    public static Term parseTerm(String term) {
        String[] termParts = term.trim().split(" ");
        return new Term(Season.getSeason(termParts[0]), Integer.valueOf(termParts[1]));
    }

    /**
     * Gets the Term today falls in
     *
     * @return Today's corresponding currentTerm
     */
    @NonNull
    public static Term currentTerm() {
        LocalDate today = LocalDate.now();
        int month = today.getMonthValue();
        int year = today.getYear();

        if (month >= 9 && month <= 12) {
            return new Term(Season.FALL, year);
        } else if (month >= 1 && month <= 4) {
            return new Term(Season.WINTER, year);
        } else {
            return new Term(Season.SUMMER, year);
        }
    }
}

File: app/src/main/java/com/guerinet/mymartlet/model/Term.kt
/*
 * Copyright 2014-2018 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.model

import android.content.Context

import org.threeten.bp.LocalDate

/**
 * One class term, consisting of a season and a year
 * @author Julien Guerinet
 * @since 1.0.0
 *
 * @param season    Term [Season]
 * @param year      Term year
 */
@Suppress("EqualsOrHashCode")
class Term(val season: Season, val year: Int) {

    /**
     * Term Id, for parsing errors
     */
    val id = "${season.title} $year"

    /**
     * Returns true if the current term if after the given [term], false otherwise
     */
    fun isAfter(term: Term): Boolean = when {
        // Year after
        year > term.year -> true
        // Year Before
        year < term.year -> false
        // Same year: check the semesters
        else -> season.number.toInt() > term.season.number.toInt()
    }

    /**
     * Returns the String representation of this term, using the [context]
     */
    fun getString(context: Context): String = "${season.getString(context)} $year"

    /**
     * Returns the term in the format used by McGill
     */
    override fun toString(): String = year.toString() + season.number

    override fun equals(other: Any?): Boolean {
        if (other !is Term) {
            return false
        }
        return season == other.season && year == other.year
    }

    companion object {

        /**
         * Returns the term from the [term] in the MGill String format (Ex: 199901 is Winter 1999)
         */
        fun parseMcGillTerm(term: String): Term {
            // Split it into the year and the season
            val year = Integer.parseInt(term.substring(0, 4))
            val season = Season.getSeasonFromNumber(term.substring(4))
            return Term(season, year)
        }

        /**
         * Returns a parsed term from the [term] String
         */
        fun parseTerm(term: String): Term {
            val termParts = term.trim().split(" ")
            return Term(Season.getSeasonFromTitle(termParts[0]), termParts[1].toInt())
        }

        /**
         * Returns today's corresponding term
         */
        fun currentTerm(): Term {
            val today = LocalDate.now()
            val year = today.year

            return when (today.monthValue) {
                in 9..12 -> Term(Season.FALL, year)
                in 1..4 -> Term(Season.WINTER, year)
                else -> Term(Season.SUMMER, year)
            }
        }
    }
}
